secure/storage/github/src/lib.rs

```diff
@@ -0,0 +1,289 @@
+    fn get_internal(&self, path: &str) -> Result<Vec<GetResponse>, Error> {
+        let resp = self.upgrade_request(ureq::get(&self.url(path))).call();
+        match resp.status() {
+            200 => {
+                let resp = resp.into_string()?;
+                let get_resp: Result<GetResponse, Error> =
+                    serde_json::from_str(&resp).map_err(|e| e.into());
+
+                if let Ok(get_resp) = get_resp {
+                    return Ok(vec![get_resp]);
+                }
+
+                let get_resp: Result<Vec<GetResponse>, Error> =
+                    serde_json::from_str(&resp).map_err(|e| e.into());
+                if let Ok(get_resp) = get_resp {
+                    return Ok(get_resp);
+                }
+
+                Err(Error::SerializationError(resp))
+            }
+            404 => Err(Error::NotFound(self.url(path))),
+            _ => Err(resp.into()),
+        }
+    }
+
+    /// Returns the sha hash of a file according to GitHub
+    fn get_sha(&self, path: &str) -> Result<Option<String>, Error> {
+        match self.get_internal(path) {
+            Ok(value) => {
+                if let Some(value) = value.iter().find(|entry| entry.path == path) {
+                    Ok(Some(value.sha.clone()))
+                } else {
+                    // This is a directory
+                    Err(Error::InternalError(format!(
+                        "get mismatch, found {} entries",
+                        value.len()
+                    )))
+                }
+            }
+            Err(Error::NotFound(_)) => Ok(None),
+            Err(e) => Err(e),
+        }
+    }
+
+    fn url(&self, path: &str) -> String {
+        format!(
+            "{}/repos/{}/{}/contents/{}",
+            URL, self.owner, self.repository, path
+        )
+    }
+}
+
+#[derive(Debug, Deserialize, PartialEq, Serialize)]
+struct GetResponse {
+    #[serde(rename = "type")]
+    content_type: String,
+    path: String,
+    sha: String,
+    content: Option<String>,
+}
+
+// This test suite depends on the developer providing an active repository, the account where that
+// repository is hosted, and a token that can modify that repository. Due to the nature of these
+// tests, they cannot be provided in this framework and the tests are labeled as ignored
+//
+// GitHub updates can misbehave due to changes in file structure thus dependent operations can
+// fail, for example, an update of a file reads its hash and then updates. If another operation
+// happens after the hash retrieval but before the update, the hash will change resulting in a
+// failed update. Therefore these tests must be run in series via:
+// `cargo xtest -- --ignored --test-threads=1`
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    const OWNER: &str = "OWNER";
+    const REPOSITORY: &str = "REPOSITORY";
+    const TOKEN: &str = "TOKEN";
+
+    #[ignore]
+    #[test]
+    fn test_files() {
+        let path = "data";
```

Comment:
nit: I'd make some of these more obvious, i.e.,
let path = "data.txt" (so I know it's a file)
let value1 = "data_contents_1"
let value2 = "data_contents_2"

Reply:
very much a @JoshLind thing ;)